### DISCUSSION TITLE
Minimum number of wires in cable running between hands.

### DISCUSSION THREAD

**Discussion by TactileDesign** [kriswilk/buzzah#17] [Q&A]
Thanks to everyone here for all the great information!  I am interested in mounting all of the electronics for my upcoming Buzzah  gloves so I dont have to carry around a fanny pack.  (See pic of my current gloves using ERMs).  I hope to mount 4 motor controllers on my right glove that will not have the multiplexer.  My understanding is that each controller requires 4 wires: V0, Ground, SCL,and SDA.  In order to minimize the number of wires I need to run from the LH multiplexer glove to the RH glove, I will use common V0 and Ground wires.  My question is this: Since SCL is a clock signal from the processor, Can I use a common wire to feed all 4 of the SCLs on my right glove??  If so, I can get by with 7 wires.  If each driver board needs its own SCL and SDA, then I will need 10 - too many for an RJ45 connector.  I can just use a fat 10 conductor cable but I would prefer not...  THANKS!
![gloves](https://github.com/kriswilk/buzzah/assets/172641786/0cbca0ca-f9e0-4b1b-be59-42485f169e1b)

**Comment by kriswilk** [comment]
Sorry for the delay, I just had a chance to get back to this repository and saw your message. The SCL and SDA signals are I2C bus signals that are shared by all devices on the bus. You only need one wire for each to carry the signals between gloves. So 4 wires in total should be adequate: power, ground, SCL, and SDA.

Let me know if you need any further clarification.

**Reply by TactileDesign** [reply]
Wow!  Thanks for your response.  Only 4 is great!  I have already committed to a fanny pack for my current build, but I will make use of this info in future versions...  But, if all the boards are in parallel, how does the multiplexer identify which driver board goes with which finger, or which set of 4 driver boards goes with which hand?

**Reply by kriswilk** [reply]
The ports on the mux are numbered (0-7) and hence individually addressable. The code treats (0-3) and (4-7) as the two hands. Whether you wire them 0-index to 3-pinky or the other way around isn't important since the patterns are randomized. However, if you intend to use the "mirror" feature in the code, you should make sure that ports 0/4, 1/5, 2/6, 3/7 correspond to the same fingers on each hand.

**Reply by TactileDesign** [reply]
Kris, Sorry to bother you again with seemingly the same question, but I am still unsure in regard to the SCL clock lines for each of the finger ports coming out of the MUX.  I am pretty sure that each SDA driver board input needs to be paired with its corresponding SDA MUX output since each board has the same generic address (so that's 4 SDA wires per hand). 

But, you previous response seems to suggest that I can I run all of the 2305 driver board SCL input ports in parallel with the main RP2040 SCL out which feeds the MUX.  Did I interpret that correctly?  Or, does each SCL motor driver board SCL input need to be paired with its corresponding SCL output port 0-7 on the MUX?  

Here is what the Adafruit spec sheet says:  
"There are 8 sets of SDA and SCL pins, from SD0/SC0 to SD7/SC7. These are the multiplexed pins. Each one is a completely separate I2C bus set." 
(So this leads me to think that each bus set is different, and SCL pins going out to all the driver boards cant be put in parallel to lower the number of wires I need)  

But then the spec sheet goes on to say:  
"You can have 8 I2C devices with identical addresses, as long as they are on one I2C bus each." 
(I note that the spec sheet does not say "...one 12C bus _set_ each", instead it refers to a single bus for each driver board.  So either Adafruit has a typo, and they should have referred to a unique "bus set" for each I2C device, OR the "one I2C bus each" they refer to is the SDA bus, and they think that it is not even worth mentioning that the SCL clock bus is the same parallel clock bus that is common to each driver board.)

It seems common sense that the clock lines should all be the same (as it is in digital music applications), but I wonder if the clock for a driver might only turn on when the respective driver board is called upon to operate....

**Reply by kriswilk** [reply]
Thanks for re-opening this dialogue because I was not really thinking clearly the first time around! Maybe I'll start by backing up and explaining why the mux is needed in the first place, for anyone who may read this thread in the future.

The I2C bus is a board-level interconnect that allows many different devices to be connected to one pair of active pins (SDA, SCL) on a microcontroller. All the devices on the bus have unique 8-bit "addresses", and each device listens for instructions from the processor that are directed at its own address.

Problems arise when an application requires two or more identical devices (with the same address). This is the situation with the motor drivers: unlike some I2C devices, their addresses cannot be changed.

The solution is the multiplexer: it has its own address, and accepts instructions that include the port number, address, and payload to deliver there. The mux has 8 independent bus channels, which can each have many devices chained off of them. In effect, the motor drivers are not connected to the processor at all...instructions are merely handed off to each driver by the mux as they arrive.

OK, so back to your original question... if the mux is in a central unit and you want to run wires out to gloves containing the drivers/motors, then my original answer is indeed WRONG. You will need separate sets of wires for each driver, since each is on a separate bus. Well, technically the power and ground can be shared. But the SCL/SDA for each driver should be independent.

That said (you'll have seen this coming if you read my comment in the other thread today), I'd advise against separating the mux and drivers by more than a few inches. Longer than this and the signals may degrade to the point that you get glitches and errors. The I2C bus was designed to connect chips on a single PCB, not widely-separated devices.

**Reply by TactileDesign** [reply]
Wow!  It is amazing such small lengths of wire could have enough capacitance to cause problems - we have to be dealing with magnitudes in tenths of picofarads....   My newest design for my 2nd gen set up calls for a run of just over 6 inches.  Adafruit says here: https://learn.adafruit.com/working-with-i2c-devices/cable-length
that I should be OK if our bus is 400 khz or less.  What frequency we are running our bus at?

**Reply by kriswilk** [reply]
CircuitPython [defaults to 100 kHz](https://docs.circuitpython.org/en/latest/shared-bindings/busio/#busio.I2C:~:text=frequency%3A%20int%20%3D%20100000), so the rule of thumb would be about 12" for the maximum point-to-point length on each bus.
